replaydetector/replaydetector_test.go

```diff
@@ -84,74 +102,296 @@ func TestReplayDetector(t *testing.T) {
-			nil,
 		},
 		"ContinuousReplayed": {
 			8, 0x0000FFFFFFFFFFFF,
 			[]uint64{16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25},
 			[]bool{
 				true, true, true, true, true, true, true, true, true, true,
+				false, false, false, false, false, false, false, false, false, false,
+			},
+			[]bool{
 				true, true, true, true, true, true, true, true, true, true,
+				false, false, false, false, false, false, false, false, false, false,
 			},
 			[]uint64{16, 17, 18, 19, 20, 21, 22, 23, 24, 25},
-			nil,
 		},
 		"ReplayedLater": {
 			128, 0x0000FFFFFFFFFFFF,
 			[]uint64{16, 32, 48, 64, 80, 96, 112, 128, 16, 32, 48, 64, 80, 96, 112, 128},
 			[]bool{
-				true, true, true, true, true, true, true, true, true, true,
-				true, true, true, true, true, true,
+				true, true, true, true, true, true, true, true, false, false,
+				false, false, false, false, false, false,
+			},
+			[]bool{
+				true, true, true, true, true, true, true, true, false, false,
+				false, false, false, false, false, false,
 			},
 			[]uint64{16, 32, 48, 64, 80, 96, 112, 128},
-			nil,
 		},
 		"ReplayedQuick": {
 			128, 0x0000FFFFFFFFFFFF,
 			[]uint64{16, 16, 32, 32, 48, 48, 64, 64, 80, 80, 96, 96, 112, 112, 128, 128},
+			[]bool{
+				true, false, true, false, true, false, true, false, true, false,
+				true, false, true, false, true, false,
+			},
+			[]bool{
+				true, false, true, false, true, false, true, false, true, false,
+				true, false, true, false, true, false,
+			},
+			[]uint64{16, 32, 48, 64, 80, 96, 112, 128},
+		},
+		"Strict": {
+			0, 0x0000FFFFFFFFFFFF,
+			[]uint64{1, 3, 2, 4, 5, 6, 7, 8, 9, 10},
+			[]bool{
+				true, true, false, true, true, true, true, true, true, true,
+			},
+			[]bool{
+				true, true, false, true, true, true, true, true, true, true,
+			},
+			[]uint64{1, 3, 4, 5, 6, 7, 8, 9, 10},
+		},
+		"Overflow": {
+			128, 0x0000FFFFFFFFFFFF,
+			[]uint64{0x0000FFFFFFFFFFFE, 0x0000FFFFFFFFFFFF, 0x0001000000000000, 0x0001000000000001},
+			[]bool{
+				true, true, false, false,
+			},
+			[]bool{
+				true, true, false, false,
+			},
+			[]uint64{0x0000FFFFFFFFFFFE, 0x0000FFFFFFFFFFFF},
+		},
+	}
+	for name, c := range cases {
+		c := c
+		t.Run(name, func(t *testing.T) {
+			det := New(c.windowSize, c.maxSeq)
+			var out []uint64
+			for i, seq := range c.input {
+				accept, ok := det.Check(seq)
+				if ok != c.valid[i] {
+					t.Errorf("Unexpected validity (%d):\nexpected: %v\ngot: %v", seq, c.valid[i], ok)
+				}
+				if ok {
+					out = append(out, seq)
+				}
+				if latest := accept(); latest != c.latest[i] {
+					t.Errorf("Unexpected sequence latest status (%d):\nexpected: %v\ngot: %v", seq, c.latest[i], latest)
+				}
+			}
+			if !reflect.DeepEqual(c.expected, out) {
+				t.Errorf("Wrong replay detection result:\nexpected: %v\ngot:      %v",
+					c.expected, out,
+				)
+			}
+		})
+	}
+}
+
+func TestReplayDetectorWrapped(t *testing.T) {
```

Comment:
Is there room to refactor these two top level tests such that the cases are the same but whether we wrap or not is different? Might save a decent amount of lines from being added. If not, all good!

Reply:
@edaniels potentially -- they were structured in that manner prior to my changes here. My reasoning for breaking them out was that when adding the check for whether the `latest` was updated on a given input, the test cases ended up with `expected` / `expectedWrap` (which was already present), `valid` / `validWrap`, and `latest` / `latestWrap`, at which point it started to feel like we were trying to run two separate sets of tests in the same test cases. One of the nice attributes of table-driven tests is being able to visually look at a test case and see what behavior it is intended to exercise, which felt like was getting a bit lost (this was probably exacerbated by the wrapped output defaulting to the non-wrapped output when the wrapped was `nil`). The downside of my approach is increase in lines added (as you mentioned) as well as potential lack of maintainability of the two separate sets of cases. I think I personally lean towards the solution I went with here, but I don't feel too strongly. One middle ground option would be to combine them again, but have something like:

```
map[string]struct {
		windowSize uint
		maxSeq     uint64
		input      []uint64
		expected map[string]struct{
			valid      []bool
			latest     []bool
			expected   []uint64
		}
	}
```

Thoughts?

Reply:
Ah yes that could work. I think what I was going for was something like this where we reuse the same cases but extend them where needed:
```golang
type testCase struct {
	windowSize uint
	maxSeq     uint64
	input      []uint64
	valid      []bool
	latest     []bool
	expected   []uint64
}

const (
	largeSeq = 0x100000000000
	hugeSeq  = 0x1000000000000
)

var commonTestCases = map[string]testCase{
	"Continuous": {
		16, 0x0000FFFFFFFFFFFF,
		[]uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20},
		[]bool{
			true, true, true, true, true, true, true, true, true, true,
			true, true, true, true, true, true, true, true, true, true,
			true,
		},
		[]bool{
			true, true, true, true, true, true, true, true, true, true,
			true, true, true, true, true, true, true, true, true, true,
			true,
		},
		[]uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20},
	},
}

func checkInput(t *testing.T, det ReplayDetector, c testCase) {
	var out []uint64
	for i, seq := range c.input {
		accept, ok := det.Check(seq)
		if ok != c.valid[i] {
			t.Errorf("Unexpected validity (%d):\nexpected: %v\ngot: %v", seq, c.valid[i], ok)
		}
		if ok {
			out = append(out, seq)
		}
		if latest := accept(); latest != c.latest[i] {
			t.Errorf("Unexpected sequence latest status (%d):\nexpected: %v\ngot: %v", seq, c.latest[i], latest)
		}
	}
	if !reflect.DeepEqual(c.expected, out) {
		t.Errorf("Wrong replay detection result:\nexpected: %v\ngot:      %v",
			c.expected, out,
		)
	}
}

func TestReplayDetector(t *testing.T) {
	for name, c := range commonTestCases {
		c := c
		t.Run(name, func(t *testing.T) {
			det := New(c.windowSize, c.maxSeq)
			checkInput(t, det, c)
		})
	}
}

func TestReplayDetectorWrapped(t *testing.T) {
	wrappedCases := map[string]testCase{
		"WrapContinuous": {
			64, 0xFFFF,
			[]uint64{0xFFFC, 0xFFFD, 0xFFFE, 0xFFFF, 0x0000, 0x0001, 0x0002, 0x0003},
			[]bool{
				true, true, true, true, true, true, true, true,
			},
			[]bool{
				true, true, true, true, true, true, true, true,
			},
			[]uint64{0xFFFC, 0xFFFD, 0xFFFE, 0xFFFF, 0x0000, 0x0001, 0x0002, 0x0003},
		},
	}
	for name, c := range commonTestCases {
		if _, ok := wrappedCases[name]; ok {
			t.Fatalf("duplicate test: %q", name)
		}
		wrappedCases[name] = c
	}
	for name, c := range wrappedCases {
		c := c
		t.Run(name, func(t *testing.T) {
			det := WithWrap(c.windowSize, c.maxSeq)
			checkInput(t, det, c)
		})
	}
}
```

What do you think of that?

Reply:
@edaniels ah, the issue with doing as you suggest is that the `valid` and `latest` are not the same for every input across wrapped / not wrapped. This is what led me to split them in the first place. A good example is the following (let's say window of `20`:
- `1` (not wrapped: `valid: true` / `latest: true`, wrapped: `valid: true` / `latest: true`)
- `2` (not wrapped: `valid: true` / `latest: true`, wrapped: `valid: true` / `latest: true`)
- `67` (not wrapped: `valid: true` / `latest: true`, wrapped: `valid: true` / `latest: true`)
- `5` (not wrapped: `valid: false` / `latest: false`, wrapped: `valid: true` / `latest: true`)
- `3` (not wrapped: `valid: false` / `latest: false`, wrapped: `valid: true` / `latest: false`)

Reply:
so when mocking this out, I was looking at which test cases are 100% and it's every single case except for the new names introduced in the wrapped variants. Well rather, all the test cases passed so I assume they are equivalent :P

Reply:
Ah fair point -- I'll pull those out 👍🏻

Reply:
@edaniels updated 👍🏻 thanks for the feedback!